### PR TITLE
MINOR: Improve range checking in MockLog

### PR DESCRIPTION
**Pull request by hachikuji**
This patch includes a few small fixes in `MockLog`. In particular, it makes the range checking consistent with the standard `Log` implementation. It also adds some additional testing and fixes some of the start offset return logic in `read`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/test/java/org/apache/kafka/raft/MockLog.java [hunk 1]
```diff
@@ -220,46 +230,68 @@ public LogAppendInfo appendAsFollower(Records records) {
...
+        verifyOffsetInRange(startOffset);
 
+        long maxOffset = maxOffsetOpt.orElse(endOffset().offset);
         if (startOffset == maxOffset) {
             return Collections.emptyList();
         }
 
         return log.stream()
-            .filter(batch -> batch.firstOffset() >= startOffset && batch.lastOffset() < maxOffset)
+            .filter(batch -> batch.lastOffset() >= startOffset && batch.lastOffset() < maxOffset)
```
Comment (guozhangwang) [inline raft/src/test/java/org/apache/kafka/raft/MockLog.java:241]: Is this intentional?

File: raft/src/test/java/org/apache/kafka/raft/MockLog.java [hunk 2]
```diff
@@ -220,46 +230,68 @@ public LogAppendInfo appendAsFollower(Records records) {
...
-            if (batch.lastOffset() >= startOffset && batch.lastOffset() < maxOffset) {
-                buffer = batch.writeTo(buffer);
+            if (batch.lastOffset() >= startOffset) {
+                if (batch.lastOffset() < maxOffset) {
+                    buffer = batch.writeTo(buffer);
+                }
 
-                if (startId == null && !batch.entries.isEmpty()) {
-                    startId = batch.entries.iterator().next().id;
+                if (firstEntry == null && !batch.entries.isEmpty()) {
```
Comment (guozhangwang) [inline raft/src/test/java/org/apache/kafka/raft/MockLog.java:280]: Nice one.

File: raft/src/test/java/org/apache/kafka/raft/MockLog.java (hunk 1 above)
Comment (hachikuji) [reply raft/src/test/java/org/apache/kafka/raft/MockLog.java:241]: Yeah, I think this was a bug. We need to include the batch if it contains `startOffset`. The previous check only worked when the fetch offset is aligned at the start of the batch. This is consistent with the logic in `read`.

Comment (guozhangwang) [reply raft/src/test/java/org/apache/kafka/raft/MockLog.java:241]: Acked the `Log` and confirmed, I was originally thinking that if a fetch start offset is in the middle of a batch (in practice it should not happen at least from our own Java clients though) we would start from the next batch. Realized it's not the case actually :)

**Pull request merged**